### ISSUE TITLE
A few issues I've noticed so far...

### ISSUE THREAD

**Issue by mattwarkentin** [mlverse/luz#4]
Hi @dfalbel,

Really cool new package. I look forward to it maturing. A few things I've noticed so far when trying to run some of the examples:

- I think you need to re-export `fit`, since this function is not loaded by `luz`, you end up getting the following error, which goes away if you manually call `library(generics)`
  ```
  Error in fit(., train_dl, epochs = 10, valid_data = test_dl) : 
    could not find function "fit"
  ```

  Something like this should do the trick, I think
  ```r
  #' @importFrom generics fit
  #' @export
  generics::fit
  ```

- None of the metrics or callbacks are exported, so running the code below also triggers an error
  ```r
  fitted <- net %>%
    setup(
      loss = nn_cross_entropy_loss(),
      optimizer = optim_adam,
      metrics = list(
        luz_metric_accuracy
      )
    ) %>%
    fit(train_dl, epochs = 10, valid_data = test_dl)
  ```
  ```
  Error in setup(., loss = nn_bce_with_logits_loss(), optimizer = optim_adam,  : 
    object 'luz_metric_accuracy' not found
  ```

**Comment by mattwarkentin** [comment]
Also for the examples to be generally self-contained example they should probably also load `luz` at the beginning:
````

**Comment by dfalbel** [comment]
Thanks very much for the valuable feedback @mattwarkentin , will fix this soon

**Comment by mattwarkentin** [comment]
Sorry I would've contributed a PR but I have been very busy this week :(

**Comment by mattwarkentin** [comment]
Would it be possible to extend `luz_callback_progress()` so that the user could pass in a custom format string (i.e. `":current/:total [:bar] - ETA: :eta"`)?
https://github.com/mlverse/luz/blob/e8fd4dcdea6a18ac9514dee12a11f6bac0cd13ca/R/callbacks.R#L73-L81

**Comment by dfalbel** [comment]
My main concern is that it's really easy break the callback specially because of the strong interaction with the metrics callback. 

We could use the progressr package that uses these `handler` (which seems to be basically an `option`) to control the format. Since it allows printing messages, we would still be able to print metrics and loss values during the iterations.
The main drawback is that we no longer control the defaults, so for example the default wouldn't have ETA, or iter number , etc.. (and also the user would need to explicitly request the progressbar with `with_progress`).

I'd rather have a consensus on what we prefer to have in the by default than allowing too much customizations, for  example, do you think something like: `:epoch/:epochs :current/:total [:bar] - ETA: :eta - Ellapsed: elapsedfull`. 
Ultimately, the user could set `verbose=FALSE` and write their own specific callback, or we could even leave this as the default but provide a another `luz_callback_progressr`?

**Comment by mattwarkentin** [comment]
Sorry for the slow response.

> My main concern is that it's really easy break the callback specially because of the strong interaction with the metrics callback.

This is fair. And in a high-level API like `luz`, it probably makes sense to have an opinionated approach.

> We could use the progressr package...and also the user would need to explicitly request the progressbar with with_progress

Yes, this is exactly what I was thinking originally. `progressr` now supports `handlers(global = TRUE)` which removes the need to wrap any expressions in `with_progress()` (see https://github.com/HenrikBengtsson/progressr/issues/95).

I did try to implement a `progressr` that played nicely with the current callbacks but I ran into an issue that I am trying to sort out here https://github.com/HenrikBengtsson/progressr/issues/111.

> I'd rather have a consensus on what we prefer to have in the by default than allowing too much customizations

I agree. Upon reflection, I don't know that `:eta` and `:elapsed` are of interest to most people. I think my preference is probably: `:epoch/:epochs :current/:total [:bar] - ETA: :eta`

**Comment by dfalbel** [comment]
Cool I'll make sure we have `:epoch/:epochs` in the progress bar by default too. 
Regarding https://github.com/HenrikBengtsson/progressr/issues/111 I think it would be already nice if the user could control how they prefer the progress bar (with the defaults) and we handle the message (which is basically pasting the metrics and their values) from luz side. The displayed metrics can be somewhat easily controlled from luz side by changing the tracked metrics.

**Comment by dfalbel** [comment]
Regarding elapsed, I think it would be nice to implement a callback that times the higher level operations in the luz loop. Similar to:

https://pytorch-lightning.readthedocs.io/en/stable/advanced/profiler.html

THis would allow us to quickly find what are the bottlenecks.

**Comment by mattwarkentin** [comment]
If constructing the message is handled on the `luz` side of things, then you're right, the user can just set their preferred progress handler(s) and enable global progress handling and the rest would be handled by `luz`. This would be pretty straightforward actually.

> Regarding elapsed, I think it would be nice to implement a callback that times the higher level operations in the luz loop.

Agreed! A profiler callback would be an awesome addition.

**Comment by dfalbel** [comment]
Yeah, I think the only thing is that we would be able to have `epochs` in the beginning of the progress bar :( because they would be custom tokens...

**Comment by mattwarkentin** [comment]
If you want to implement a `progressr` callback, I wouldn't format the message to work with `progress` progress bars, as this won't work for any other progress handler. The goal of `progressr` is to make it easy for users to choose their favourite way of getting progress updates. Also, I don't think it makes much sense to implement a `progressr` callback gear towards `progress`, as there is already be a `progress` callback.

The message should be something generic/universal  that could be used by any of the currently available progress handlers (and those added in the future). In other words, the message should be constructed independently from what handler someone might use. Something like `"Epoch {epoch} - Metric1: {value1}, Metric 2: {value2}"`.

The code below shows a toy example of the "developer" constructing the message, but the user choosing/trying different handlers.
```r
library(progressr)
library(glue)

handlers(global = TRUE)

slow_fun <- function() {
  p <- progressor(100)
  for (i in 1:100) {
    Sys.sleep(0.1)
    p(glue("Epoch {i} - Metric: {runif(1)}"))
  }
}

handlers(
  handler_progress()
)

slow_fun()

handlers(
  handler_pbcol()
)

slow_fun()

handlers(
  handler_rstudio()
)

slow_fun()
```

**Comment by dfalbel** [comment]
Yes! This makes sense to me! I am convinced that this is better than the current strategy we have. Do you want to implement this functionality?

**Comment by mattwarkentin** [comment]
Sure, I'd be happy to contribute.

**Comment by mattwarkentin** [comment]
On a side note, for `luz_callback_progress()`, I don't know that epoch is needed in the format string (`":epoch/:epochs :current/:total [:bar] - ETA: :eta"`) if the epoch is also still printed out `on_epoch_end`
https://github.com/mlverse/luz/blob/24284fd2f6b2b813d618f3498ac2a750c8ada979/R/callbacks.R#L98-L104

Probably one or the other, to avoid duplicate information. What do you think?

**Comment by dfalbel** [comment]
I agree. I don't have a strong opinion on any of them though. The difference is that the progress bar disappears after training while the other is kept in the console, so we would need to print it anyways at some point.

**Comment by dfalbel** [comment]
Hey @mattwarkentin I have opened #32 and will close this one so the title is more descriptive of what we want to implement :)
